bokeh/models/formatters.py

```diff
@@ -45,6 +45,127 @@ class BasicTickFormatter(TickFormatter):
+    The numer format, as defined in the following tables:
+
+    * numbers:
+
+    +------------+--------------+---------------+
+    | Number     | Format       | String        |
+    +============+==============+===============+
+    | 10000      | '0,0.0000'   | 10,000.0000   |
+    | 10000.23   | '0,0'        | 10,000        |
+    | 10000.23   | '+0,0'       | +10,000       |
+    | -10000     | '0,0.0'      | -10,000.0     |
+    | 10000.1234 | '0.000'      | 10000.123     |
+    | 10000.1234 | '0[.]00000'  | 10000.12340   |
+    | -10000     | '(0,0.0000)' | (10,000.0000) |
+    | -0.23      | '.00'        | -.23          |
+    | -0.23      | '(.00)'      | (.23)         |
+    | 0.23       | '0.00000'    | 0.23000       |
+    | 0.23       | '0.0[0000]'  | 0.23          |
+    | 1230974    | '0.0a'       | 1.2m          |
+    | 1460       | '0 a'        | 1 k           |
+    | -104000    | '0a'         | -104k         |
+    | 1          | '0o'         | 1st           |
+    | 52         | '0o'         | 52nd          |
+    | 23         | '0o'         | 23rd          |
+    | 100        | '0o'         | 100th         |
+    +------------+--------------+---------------+
+
+    * currency:
+
+    +------------+--------------+---------------+
+    | Number     | Format       | String        |
+    +============+==============+===============+
+    | 1000.234   | '$0,0.00'    | $1,000.23     |
+    | 1000.2     | '0,0[.]00 $' | 1,000.20 $    |
+    | 1001       | '$ 0,0[.]00' | $ 1,001       |
+    | -1000.234  | '($0,0)'     | ($1,000)      |
+    | -1000.234  | '$0.00'      | -$1000.23     |
+    | 1230974    | '($ 0.00 a)' | $ 1.23 m      |
+    +------------+--------------+---------------+
+
+    * bytes:
+
+    +---------------+-----------+---------------+
+    | Number        | Format    | String        |
+    +===============+===========+===============+
+    | 100           | '0b'      | 100B          |
+    | 2048          | '0 b'     | 2 KB          |
+    | 7884486213    | '0.0b'    | 7.3GB         |
+    | 3467479682787 | '0.000 b' | 3.154 TB      |
+    +---------------+-----------+---------------+
+
+    * percentages:
+
+    +-------------+-------------+---------------+
+    | Number      | Format      | String        |
+    +=============+=============+===============+
+    | 1           | '0%'        | 100%          |
+    | 0.974878234 | '0.000%'    | 97.488%       |
+    | -0.43       | '0 %'       | -43 %         |
+    | 0.43        | '(0.000 %)' | 43.000 %      |
+    +-------------+-------------+---------------+
+
+    * time:
+
+    +------------+--------------+---------------+
+    | Number     | Format       | String        |
+    +============+==============+===============+
+    | 25         | '00:00:00'   | 0:00:25       |
+    | 238        | '00:00:00'   | 0:03:58       |
+    | 63846      | '00:00:00'   | 17:44:06      |
+    +------------+--------------+---------------+
+    """)
```

Comment:
@bryevdv, what do you mean by "engineering notation"? There was a problem (not necessarily an issue started for this), that `numeral.js` doesn't support scientific notation. I didn't mange to implement this yet, because `numeral.js` has horrible implementation. I have some nice ideas, e.g. with adding many new formats, utilizing unicode fonts (e.g. for super scripts), but most likely I will have to do major changes to `numeral.js` to get around current limitations.